src/sasl/cyrsasl_scram.erl

```diff
@@ -58,164 +58,194 @@ mech_new(_Host, Creds) ->
-                                                  R = [{username, State#state.username},
-                                                       {sasl_success_response,
-                                                        <<"v=", (jlib:encode_base64(ServerSignature))/binary>>},
-                                                       {auth_module, State#state.username}],
-                                                  {ok, mongoose_credentials:extend(State#state.creds, R)};
-                                              true -> {error, <<"bad-auth">>}
-                                           end;
-                                       _Else -> {error, <<"bad-protocol">>}
-                                   end;
-                               {$r, _} -> {error, <<"bad-nonce">>};
-                               _Else -> {error, <<"bad-protocol">>}
-                           end;
-                       true -> {error, <<"bad-channel-binding">>}
-                    end;
-                _Else -> {error, <<"bad-protocol">>}
+    ClientInList = re:split(ClientIn, <<",">>),
+    ClientInSteps = [ fun parse_step4_client_in/1,
+                      fun parse_channel_binding_attribute/1],
+    case parse_attributes(ClientInList, ClientInSteps) of
+        {ok, {NonceAtt, ClientProofAtt}} ->
+            Nonce = <<(State#state.client_nonce)/binary,
+                      (State#state.server_nonce)/binary>>,
+            {PStart, _} = binary:match(ClientIn, <<",p=">>),
+            AuthMessage = iolist_to_binary( [State#state.auth_message,",",
+                                    binary:part(ClientIn, 0, PStart)]),
+            AuthSteps = [fun verify_nonce/1,
+                         fun get_client_proof/1,
+                         fun get_stored_key/1,
+                         fun verify_stored_key/1],
+            AuthArgs = [{NonceAtt, Nonce}, ClientProofAtt, AuthMessage, State],
+            case parse_attributes(AuthArgs, AuthSteps) of
+                ok ->
+                    ServerSignature =
+                        mongoose_scram:server_signature(State#state.server_key,
+                                                        AuthMessage),
+                    R = [{username, State#state.username}, {sasl_success_response,
+                         <<"v=", (jlib:encode_base64(ServerSignature))/binary>>},
+                         {auth_module, State#state.username}],
+                    {ok, mongoose_credentials:extend(State#state.creds, R)};
+                {error, Reason} ->
+                    {error, Reason}
             end;
-        _Else -> {error, <<"bad-protocol">>}
+        {error, Reason} ->
+            {error, Reason}
     end.
 
-parse_attribute(Attribute) ->
-    AttributeLen = byte_size(Attribute),
-    case AttributeLen >= 3 of
-        true ->
-            parse_attribute(Attribute, 0);
-        _ ->
-            {error, <<"bad-format attribute too short">>}
+parse_attributes(ClientInList, Steps) ->
```

Comment:
This one would be cool if it can have some specification, more important than all the others, considering the magic this one is doing 🤔 
The specification could look something like
```erlang
-spec parse_attributes( ... , fun((...) -> ...)) -> ok | {error, term()}.
```
The only problem is that the functions being called have different input types, which makes it not possible to specify a `TList`. Is it possible to unify all input types for the handlers? Otherwise just
```erlang
 -spec parse_attributes(any(), fun()) -> ok | {error, term()}.
```
🤷‍♂